src/http/server.cr

```diff
@@ -95,90 +100,140 @@ class HTTP::Server
   {% end %}
 
   @wants_close = false
+  @sockets = [] of Socket::Server
 
-  def self.new(port, &handler : Context ->)
-    new("127.0.0.1", port, &handler)
-  end
-
-  def self.new(port, handlers : Array(HTTP::Handler), &handler : Context ->)
-    new("127.0.0.1", port, handlers, &handler)
+  # Creates a new HTTP server with the given block as handler.
+  def self.new(&handler : HTTP::Handler::Proc) : self
+    new(handler)
   end
 
-  def self.new(port, handlers : Array(HTTP::Handler))
-    new("127.0.0.1", port, handlers)
+  # Creates a new HTTP server with a handler chain constructed from the *handlers*
+  # array and the given block.
+  def self.new(handlers : Array(HTTP::Handler), &handler : HTTP::Handler::Proc) : self
+    new(HTTP::Server.build_middleware(handlers, handler))
   end
 
-  def self.new(port, handler)
-    new("127.0.0.1", port, handler)
+  # Creates a new HTTP server with the *handlers* array as handler chain.
+  def self.new(handlers : Array(HTTP::Handler)) : self
+    new(HTTP::Server.build_middleware(handlers))
   end
 
-  def initialize(@host : String, @port : Int32, &handler : Context ->)
+  # Creates a new HTTP server with the given *handler*.
+  def initialize(handler : HTTP::Handler | HTTP::Handler::Proc)
     @processor = RequestProcessor.new(handler)
   end
 
-  def initialize(@host : String, @port : Int32, handlers : Array(HTTP::Handler), &handler : Context ->)
-    handler = HTTP::Server.build_middleware handlers, handler
-    @processor = RequestProcessor.new(handler)
-  end
+  # Creates a `TCPServer` listenting on `host:port` and adds it as a socket, returning the local address
+  # and port the server listens on.
+  #
+  # If *reuse_port* is `true`, it enables the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind_tcp(host : String, port : Int32, reuse_port : Bool = false) : Socket::IPAddress
+    tcp_server = TCPServer.new(host, port, reuse_port: reuse_port)
 
-  def initialize(@host : String, @port : Int32, handlers : Array(HTTP::Handler))
-    handler = HTTP::Server.build_middleware handlers
-    @processor = RequestProcessor.new(handler)
+    bind(tcp_server)
+
+    tcp_server.local_address
   end
 
-  def initialize(@host : String, @port : Int32, handler : HTTP::Handler | HTTP::Handler::Proc)
-    @processor = RequestProcessor.new(handler)
+  # Creates a `TCPServer` listenting on `127.0.0.1:port` and adds it as a socket,
+  # returning the local address and port the server listens on.
+  #
+  # If *reuse_port* is `true`, it enables the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind_tcp(port : Int32, reuse_port : Bool = false) : Socket::IPAddress
```

Comment:
Why return the address?

Reply:
To get the actual address including the automatically assigned port when `port == 0`.

Reply:
While there is also `bind_unused_port`, it can't be used if `port` is configurable (at least not directly).